protocols/v2/roles-logic-sv2/src/vardiff/classic.rs

```diff
@@ -0,0 +1,164 @@
+use crate::utils::{hash_rate_from_target, hash_rate_to_target};
+use mining_sv2::Target;
+
+use super::Vardiff;
+
+#[allow(warnings)]
+#[derive(Debug)]
+pub struct VardiffState {
+    pub estimated_channel_hashrate: f32,
+    pub shares_per_minute: f32,
+    pub shares_since_last_update: u32,
+    pub timestamp_of_last_update: u64,
+    pub current_miner_target: Target,
+}
+
+impl VardiffState {
+    pub fn new(shares_per_minute: f32, estimated_channel_hashrate: f32) -> Self {
+        let current_miner_target =
+            hash_rate_to_target(estimated_channel_hashrate as f64, shares_per_minute as f64)
+                .expect("Cannot convert hash rate to target")
+                .into();
+        let timestamp_secs = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("time went backwards")
+            .as_secs();
+        VardiffState {
+            estimated_channel_hashrate,
+            shares_per_minute,
+            shares_since_last_update: 0,
+            timestamp_of_last_update: timestamp_secs,
+            current_miner_target,
+        }
+    }
+
+    pub fn set_shares_per_minute(&mut self, shares_per_minute: f32) {
+        self.shares_per_minute = shares_per_minute;
+    }
+
+    pub fn set_timestamp_of_last_update(&mut self, timestamp_of_last_update: u64) {
+        self.timestamp_of_last_update = timestamp_of_last_update;
+    }
+
+    pub fn set_shares_since_last_update(&mut self, shares_since_last_update: u32) {
+        self.shares_since_last_update = shares_since_last_update;
+    }
+
+    pub fn set_current_miner_target(&mut self, current_miner_target: Target) {
+        self.current_miner_target = current_miner_target;
+    }
+}
+
+impl Vardiff for VardiffState {
+    fn hashrate(&self) -> f32 {
+        self.estimated_channel_hashrate
+    }
+
+    fn shares_per_minute(&self) -> f32 {
+        self.shares_per_minute
+    }
+
+    fn last_update_timestamp(&self) -> u64 {
+        self.timestamp_of_last_update
+    }
+
+    fn shares_since_last_update(&self) -> u32 {
+        self.shares_since_last_update
+    }
+
+    fn target(&self) -> Target {
+        self.current_miner_target.clone()
+    }
+
+    fn set_hashrate(&mut self, estimated_channel_hashrate: f32) {
+        self.estimated_channel_hashrate = estimated_channel_hashrate;
+        let current_miner_target = hash_rate_to_target(
+            estimated_channel_hashrate as f64,
+            self.shares_per_minute as f64,
+        )
+        .expect("Cannot convert hash rate to targe")
```

Comment:
nit: just a typo here
```suggestion
        .expect("Cannot convert hash rate to target")
```